windows/src/desktop/kmshell/install/UfrmInstallKeyboardLanguage.pas

```diff
@@ -409,34 +411,63 @@ procedure TfrmInstallKeyboardLanguage.FillLanguageGrid;
-          FCustomLanguage.ScriptName := '';
-        if not TLanguageCodeUtils.BCP47Regions.TryGetValue(Region, FCustomLanguage.RegionName) then
-          FCustomLanguage.RegionName := '';
-        FCustomLanguage.FullName := TLanguageCodeUtils.LanguageName(FCustomLanguage.LanguageName,
-          FCustomLanguage.ScriptName, '');
-        AddRow(
-          False,
-          FCustomLanguage.FullName,
-          FCustomLanguage.FullName,
-          FCustomLanguage.ScriptName,
-          FCustomLanguage.Tag, nil);
-      finally
-        Free;
+    if IsValidLocaleName(PChar(FText)) then
+    begin
+      // Adding custom locales supported with Win8 and later
+      FAllowableText := (kmcom as IKeymanBCP47Canonicalization).GetCanonicalTag(FText);
+      if FAllowableText <> '' then
+        with TBCP47Tag.Create(FAllowableText) do
+        try
+          // Let's lookup the lang - script - region and get a good name
+          FCustomLanguage.Tag := Tag;
+          if not TLanguageCodeUtils.BCP47Languages.TryGetValue(Language, FCustomLanguage.LanguageName) then
+            FCustomLanguage.LanguageName := '';
+          if not TLanguageCodeUtils.BCP47Scripts.TryGetValue(Script, FCustomLanguage.ScriptName) then
+            FCustomLanguage.ScriptName := '';
+          if not TLanguageCodeUtils.BCP47Regions.TryGetValue(Region, FCustomLanguage.RegionName) then
+            FCustomLanguage.RegionName := '';
+          FCustomLanguage.FullName := TLanguageCodeUtils.LanguageName(FCustomLanguage.LanguageName,
+            FCustomLanguage.ScriptName, '');
+          AddRow(
+            False,
+            FCustomLanguage.FullName,
+            FCustomLanguage.FullName,
+            FCustomLanguage.ScriptName,
+            FCustomLanguage.Tag, nil);
+        finally
+          Free;
+        end;
+      FText := (kmcom as IKeymanBCP47Canonicalization).GetCanonicalTag(FText);
+      if FAllowableText <> FText then
+      begin
+        with TBCP47Tag.Create(FText) do
+        try
+          // Let's lookup the lang - script - region and get a good name
+          FCustomLanguage.Tag := Tag;
+          if not TLanguageCodeUtils.BCP47Languages.TryGetValue(Language, FCustomLanguage.LanguageName) then
+            FCustomLanguage.LanguageName := '';
+          if not TLanguageCodeUtils.BCP47Scripts.TryGetValue(Script, FCustomLanguage.ScriptName) then
+            FCustomLanguage.ScriptName := '';
+          if not TLanguageCodeUtils.BCP47Regions.TryGetValue(Region, FCustomLanguage.RegionName) then
+            FCustomLanguage.RegionName := '';
+          FCustomLanguage.FullName := TLanguageCodeUtils.LanguageName(FCustomLanguage.LanguageName,
+            FCustomLanguage.ScriptName, '');
+          AddRow(
+            False,
+            FCustomLanguage.FullName,
+            FCustomLanguage.FullName,
+            FCustomLanguage.ScriptName,
+            FCustomLanguage.Tag, nil);
```

Comment:
This block looks a little WET 😉  ll 420-441

Reply:
Fixed in https://github.com/keymanapp/keyman/pull/3557/files#diff-3a8f936acbcb68d130f21fa38f1733e8L421